### PR TITLE
Support file storage

### PR DESCRIPTION
**Pull request by juniwang**
Fix https://github.com/jenkinsci/windows-azure-storage-plugin/issues/51:  To support Azure file storage. Major changes:
- Update storage publisher(and its UI) to support both File storage and blob storage
- update builder to support file storage if "download from build" is selected.
- refactor the code structure so that upload/download codes are in separated files.
- Fix existing IT and add more file storage related IT
- reformat some files to replace tabs with spaces.

### CODE REVIEW

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 1]
```diff
@@ -158,13 +156,13 @@ public void setBlobProperties(AzureBlobProperties blobProperties) {
     }
 
     @DataBoundSetter
-    public void setCntPubAccess(boolean cntPubAccess) {
-        this.cntPubAccess = cntPubAccess;
+    public void setPubAccessible(boolean pubAccessible) {
+        this.pubAccessible = pubAccessible;
```
Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: Will there be backward compatibility issue if we rename this field?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 2]
```diff
@@ -533,9 +564,9 @@ public boolean configure(StaplerRequest req, JSONObject formData)
         /**
          * Validates storage account details.
          *
-         * @param storageAccountName
-         * @param blobEndPointURL
-         * @param storageAccountKey
+         * @param was_storageAccName
```
Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: Is there a special reason to have underscore in parameter name?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/AzureUtils.java [hunk 3]
```diff
@@ -0,0 +1,210 @@
+/*
```
Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/AzureUtils.java]: Be consistent with copyright info in other files.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/AzureUtils.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+     * @param blobURL
+     * @return
+     */
+    private static String getCustomURI(String storageAccountName, String type,
+                                       String blobURL) {
+        if (QUEUE.equalsIgnoreCase(type)) {
+            return blobURL.replace(storageAccountName + "." + BLOB,
+                    storageAccountName + "." + type);
+        } else if (TABLE.equalsIgnoreCase(type)) {
+            return blobURL.replace(storageAccountName + "." + BLOB,
```
Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/AzureUtils.java]: The logic is same as above, merge them into one.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 5]
```diff
@@ -310,53 +291,45 @@ public synchronized void perform(Run<?, ?> run, FilePath workspace, Launcher lau
...
-            if (action.getZipArchiveBlob() != null && includeArchiveZips) {
-                azureBlobs.addAll(Arrays.asList(action.getZipArchiveBlob()));
-            }
-
-            filesDownloaded = WAStorageClient.downloadArtifacts(azureBlobs, context);
-
-        } catch (WAStorageException e) {
-            context.getRun().setResult(Result.UNSTABLE);
-        }
+    private StoragePluginService getBuilderService(final BuilderServiceData data) {
```
Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: getBuilderService is not easy to understand, better to change to getDownloadService?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/service/UploadService.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+import java.io.IOException;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.StringTokenizer;
+
+public abstract class UploadService extends StoragePluginService<PublisherServiceData> {
+    protected static final String zipFolderName = "artifactsArchive";
+    protected static final String zipName = "archive.zip";
+
+    protected UploadService(PublisherServiceData serviceData) {
```
Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/service/UploadService.java]: Change PublisherServiceData -> UploadServiceData, BuilderServiceData -> DownloadServiceData to make them consistent with service class name?

File: src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/config.jelly [hunk 7]
```diff
@@ -1,94 +1,122 @@
...
+             help="/plugin/windows-azure-storage/help-storagecredential.html">
+        <c:select expressionAllowed="false"/>
+    </f:entry>
+    <f:entry title="${%filesPath_title}" field="filesPath" help="/plugin/windows-azure-storage/help-filesPath.html">
+        <f:textbox
+                checkUrl="'${rootURL}/publisher/WAStoragePublisher/checkPath?val='+encodeURIComponent(this.value)"/>
+    </f:entry>
+
+    <f:section title="${%storageType_title}">
+        <f:radioBlock name="storageType" title="Upload to Azure blob storage" value="blobstorage"
```
Comment (chenkennt) [inline src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/config.jelly]: Should this title also be in properties file?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java [hunk 8]
```diff
@@ -27,39 +29,39 @@
...
     public AzureBlobAction(Run build, String storageAccountName, String containerName,
-	    List<AzureBlob> individualBlobs, AzureBlob zipArchiveBlob,
-	    boolean allowAnonymousAccess, String storageCredentialId) {
-	this.storageAccountName = storageAccountName;
-	this.containerName = containerName;
-	this.individualBlobs = individualBlobs;
-	this.allowAnonymousAccess = allowAnonymousAccess;
-	this.zipArchiveBlob = zipArchiveBlob;
-	this.build = build;
+        List<AzureBlob> individualBlobs, AzureBlob zipArchiveBlob,
```
Comment (juniwang) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java]: updated

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java (hunk 1 above)
Comment (juniwang) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: It does have some effects but it's no that bad. For existing jobs that already built successfully,  the container was already created using old settings and won't be created again even if this field renamed. While user update the job configuration with our new version of plugin, the new field will take effect.  The only bad effect I can image is that someone create job with old version but never build successfully before upgrade to newer version. But that's not critical as well and can be fixed easily by user.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java (hunk 2 above)
Comment (juniwang) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: It's legacy code. The changes here are style-reformatting.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/AzureUtils.java (hunk 3 above)
Comment (juniwang) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/AzureUtils.java]: Updated. 

And I think we should have some talk about the license. There are several related issues.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/AzureUtils.java (hunk 4 above)
Comment (juniwang) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/AzureUtils.java]: merged

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java (hunk 5 above)
Comment (juniwang) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: `AzureStorageBuilder` extends `Builder` which is defined in Jenkins library. That's why we call it `builder` in this file. And It's not easy to rename `AzureStorageBuilder` for compatibility reason. Similarly for `WAStoragePublisher`. To make is easier to understand, I rename all new classes to `UploadXXX` or `DownloadXXX` except `AzureStorageBuilder` and `WAStoragePublisher`.

File: src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/config.jelly (hunk 7 above)
Comment (juniwang) [inline src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/config.jelly]: done

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 9]
```diff
@@ -310,53 +291,45 @@ public synchronized void perform(Run<?, ?> run, FilePath workspace, Launcher lau
...
-
-            filesDownloaded = WAStorageClient.downloadArtifacts(azureBlobs, context);
-
-        } catch (WAStorageException e) {
-            context.getRun().setResult(Result.UNSTABLE);
-        }
+    private StoragePluginService getDownloadService(final DownloadServiceData data) {
+        if (isDownloadFromContainer())
+            return new DownloadFromContainerService(data);
+        return new DownloadFromBuildService(data);
```
Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: Seems there is no place to check downloadType is 'project', add a validation?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 10]
```diff
@@ -143,10 +139,12 @@ public WAStoragePublisher(final String storageAccName,
 
     @DataBoundConstructor
     public WAStoragePublisher(final String storageCredentialId, final String filesPath,
-                              final String containerName) {
+                              final StorageType storageType) {
```
Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: This also has backward compatibility issue, let's change back to a flattened structure.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java (hunk 9 above)
Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: Do we support download from File Share here?

File: src/main/resources/com/microsoftopentechnologies/windowsazurestorage/Messages.properties [hunk 11]
```diff
@@ -16,18 +17,22 @@ WAStoragePublisher_uploaded_err=MicrosoftAzureStorage - Error occurred while upl
...
+                                          contain only letters, numbers and dashes (-). \
+                                          (Note that consecutive dashes are not allowed and dash cannot be the last character)
 WAStoragePublisher_SA_val=Validation successful
 WAStoragePublisher_uploadtype_invalid=No artifacts will be uploaded.  Either check "Upload zipped archive" or uncheck "Do not upload individual files"
 # Properties in WAStorageClient_java
 Client_SA_val_fail=Failed to validate storage account details. Please verify storage account name and key. \
-							If you are using a private or other Windows Azure cloud service, make sure that the blob endpoint url is correct.
+                            If you are using a private or other Windows Azure cloud service, make sure that the blob endpoint url is correct.
 #Properties in AzureStorageBuilder
 AzureStorageBuilder_displayName=Download from Azure Blob storage
```
Comment (chenkennt) [inline src/main/resources/com/microsoftopentechnologies/windowsazurestorage/Messages.properties]: Also remove blob here?

File: src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/config.properties [hunk 12]
```diff
@@ -1,16 +1,20 @@
 edit_title=Edit
 storageCredentials_title=Storage credentials
+storageType_title=Azure Storage type
+blobStorageType_Title=Upload to Azure blob storage
```
Comment (chenkennt) [inline src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/config.properties]: Minor: blob -> Blob

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java (hunk 9 above)
Comment (juniwang) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: Yes, it's tested.

Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: I mean download from file share seems only available when download from build? Can user directly download from a file share?

Comment (juniwang) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: No, directly from file share is not supported.  Can we add to backlog another task for this feature? to avoid this PR be expanded again.

Comment (chenkennt) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: Sure, we can do this in a separate PR.
